### PR TITLE
Add upload/download/delete/mkdir/rmdir to smb session

### PR DESCRIPTION
**Pull request by dwelch-r7**
~Merge after https://github.com/rapid7/ruby_smb/pull/262~

This PR adds the ability to upload/download/delete/mkdir/rmdir from within the smb session type

# Verification Steps
- [x] Get an smb session (enable the session type with `features set smb_session_type true)
- [x] Upload a file with `upload <local_file>` and `upload <local_file> <remote_path>`
- [x] Download a file with `download <remote_file> ` and `download <remote_file>  <local_path>`
- [x] Delete a file with `delete <file_name>
- [x] Make a directory with `mkdir <directory_name>`
- [x] Delete a directory with `rmdir <directory_name>`
- [x] Deleting a directory with files inside displays a message to the user
- [x] Verify psexec still works

### CODE REVIEW

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 1]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            @@upload_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                local_path = val
+              when 1
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_upload_help
+              end
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:367]: I assume this is missing a return

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 2]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+              print_error('No local path given')
+              return
+            end
+
+            remote_path = Rex::Post::File.basename(local_path) if remote_path.nil?
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+            fd = simple_client.open(full_path, 'wct', write: true)
+
+            data = ::File.read(local_path, ::File.size(local_path), mode: 'rb')
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:380]:
```suggestion
            data = ::File.binread(local_path)
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 3]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            fd.write(data)
+            fd.close
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:382]: Probably needs an `ensure`

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:380]: We should probably try reading the file first, so if it explodes locally - we haven't tried opening up remote connections yet - and it'll be les noise-y in the user error scenario

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 4]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            end
+
+            local_path = Rex::Post::File.basename(remote_path) if local_path.nil?
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+            fd = simple_client.open(full_path, 'o')
+            data = fd.read
+            fd.close
+
+            ::File.open(local_path, 'wb') { |local_file| local_file.write(data) }
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:430]:
```suggestion
            ::File.binwrite(local_path, data)
```

Assuming `data` is raw bytes

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 5]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            if remote_path.blank?
+              print_error('No remote path given')
+              return
+            end
+
+            local_path = Rex::Post::File.basename(remote_path) if local_path.nil?
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+            fd = simple_client.open(full_path, 'o')
+            data = fd.read
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:427]: For larger files, do we need to do chunked reads/writes? I can't remember the file size that things start to get awkward at, but potentially both my `binread`/`binwrite` API suggestion and this existing code won't support larger files that could be realistically downloaded/uploaded by users

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 6]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_upload_help
+              end
+            end
+
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+            simple_client.open(full_path, 'wc', dir: true)
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:491]: Do we need to close this resource afterwards?

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 7]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+
+            response = active_share.open_directory(directory: full_path, write: true, delete: true)
+            directory = RubySMB::SMB2::File.new(name: full_path, tree: active_share, response: response, encrypt: @tree_connect_encrypt_data)
+            # TODO: This doesn't actually delete the directory
+            # response is "{Access Denied} A process has requested access to an object but has not been granted those access rights."
+            # despite passing in `delete: true` to the `open_directory` function, not sure where the issue lies yet
+            directory.delete
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:528]: Same as above, do we need to close this ourselves

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 8]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+                cmd_delete_help
+              end
+            end
+
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+
+
+            response = active_share.open_directory(directory: full_path, write: true, delete: true)
+            directory = RubySMB::SMB2::File.new(name: full_path, tree: active_share, response: response, encrypt: @tree_connect_encrypt_data)
+            # TODO: This doesn't actually delete the directory
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:525]: This sounds like a blocker 🕵️

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 9]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+
+            @@download_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                remote_path = val
+              when 1
+                local_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_download_help
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:414]:
```suggestion
                cmd_download_help
                return
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 10]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            end
+            remote_path = nil
+
+            @@delete_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_delete_help
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:455]:
```suggestion
                cmd_delete_help
                return
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 11]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            end
+            remote_path = nil
+
+            @@mkdir_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_upload_help
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:485]:
```suggestion
                cmd_upload_help
                return
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 12]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+            end
+            remote_path = nil
+
+            @@delete_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_delete_help
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:516]:
```suggestion
                cmd_delete_help
                return
```

File: spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb [hunk 13]
```diff
@@ -13,6 +16,9 @@
   end
 
   before(:each) do
+    allow(client).to receive(:dispatcher).and_return(dispatcher)
+    allow(dispatcher).to receive(:tcp_socket).and_return(socket)
```
Comment (adfoster-r7) [inline spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb:20]:
```suggestion
```

```
  let(:simple_client) { instance_double(Rex::Proto::SMB::SimpleClient, dispatcher: dispatcher) }
  let(:dispatcher) { instance_double(RubySMB::Dispatcher::Socket, socket: socket) }
```

Not tested

File: spec/lib/rex/post/smb/ui/console/command_dispatcher/shares_spec.rb [hunk 14]
```diff
@@ -14,6 +17,9 @@
   end
 
   before(:each) do
+    allow(client).to receive(:dispatcher).and_return(dispatcher)
+    allow(dispatcher).to receive(:tcp_socket).and_return(socket)
```
Comment (adfoster-r7) [inline spec/lib/rex/post/smb/ui/console/command_dispatcher/shares_spec.rb:21]: Same as above 🤞

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 15]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
           def cmd_cd_tabs(_str, words)
             return [] if words.length > 1
 
-            @@cat_opts.option_keys
+            @@cd_opts.option_keys
+          end
+
+          def cmd_upload(*args)
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:350]: I think some of these methods are missing this one liner

```
            return print_no_share_selected unless active_share
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 16]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+          end
+
+          def cmd_upload_help
+            print_line 'Usage: upload <local_path> <remote_path>'
+            print_line
+            print_line 'Upload a file to the remote target.'
+            print @@upload_opts.usage
+          end
+
+          def cmd_download(*args)
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:398]: Placeholder comment from me incase I forgot; Looks like this doesn't work for me

This branch

```
SMB (127.0.0.1\foo_share) > ls
ls
==

    #  Type  Name  Created                    Accessed                   Written                    Changed                    Size
    -  ----  ----  -------                    --------                   -------                    -------                    ----
    0  DIR   .     2024-02-28T13:39:12+00:00  2024-02-28T13:43:59+00:00  2024-02-28T13:39:12+00:00  2024-02-28T13:39:12+00:00
    1  DIR   ..    2024-02-28T13:45:18+00:00  2024-02-28T13:45:18+00:00  2024-02-28T13:45:25+00:00  2024-02-28T13:45:25+00:00
    2  FILE  bar   2024-02-28T13:39:12+00:00  2024-02-28T13:45:25+00:00  2024-02-28T13:39:12+00:00  2024-02-28T13:39:12+00:00  4

SMB (127.0.0.1\foo_share) > download bar
[-] Error running command download: RubySMB::Error::UnexpectedStatusCode The server responded with an unexpected status code: STATUS_OBJECT_NAME_NOT_FOUND

```

Whilst with smb client:
```
smb: \> get bar  
getting file \bar of size 4 as bar (1.3 KiloBytes/sec) (average 1.3 KiloBytes/sec)
```

Comment (adfoster-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:398]: I tried taking a quick look at wireshark, but it looks like the login scanner doesn't honor the `smb::alwaysencrypt=false` and `smb::protocolversion=2` options - so it's all encrypted by default

Comment (dwelch-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:398]: that's weird, can you `cat` the file?

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 17]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
           def cmd_cd_tabs(_str, words)
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:344]: I can't comment on it; but can we update `cmd_cat` to use simple client too? The current implementation seems to fail if you cat the same empty file twice

```
SMB (127.0.0.1\foo_share) > cat foo

SMB (127.0.0.1\foo_share) > cat foo
[-] RubySMB::Error::EncryptionError Communication error with the remote host: Socket read returned nil. The server supports encryption but was not able to handle the encrypted request.
```

Whilst download works, which uses the simple wrapper

```
SMB (192.168.123.135\foo) > download foo.txt
[+] Downloaded foo.txt to foo.txt
SMB (192.168.123.135\foo) > download foo.txt
[+] Downloaded foo.txt to foo.txt
```

**Comment by cgranleese-r7** [comment]
# Tested against Windows 10

A couple of the `<cmd> -h` commands points at the incorrect help output, probably worth double checking them all. I added any I came across while testing below under the corresponding command.

## `upload`
Is the intended functionality to overwrite a file if one exists with the same name? Not familiar with how the SMB client handles that so just curious.
image

## `download`
### remote 
image

### local after downloading from session
image

### `delete` 
image

Noticed that `delete -h` returns the help output for download
image


### `mkdir`
`mkdir` also worked when passing a directory with spaces 👍 
image

`mkdir -h` returns upload help info:
image

### `rmdir`
image

`rmdir -h` returns download help info:
image


> [!Note]
Not sure if this is an issue and isn't related to this PR but thought I'd call it out. Should `ls` be taking paths and return the appropriate output for the given path?
Currently if I pass `ls` a directory it just returns the `pwd`'s `ls` output, rather than returning the directory I passed `ls` output:
image

If that is something we'd want to do, we should also update the output to have the passed path be returned somewhere like here maybe?
```
SMB (192.168.175.193\my_share\test) > ls
ls (\test) <-- Something like this
==

    #  Type  Name          Created                  Accessed       Written        Changed        Size
    -  ----  ----          -------                  --------       -------        -------        ----
    0  DIR   .             2023-12-13T15:25:18+00:  2024-03-01T11  2024-03-01T11  2024-03-01T11
                           00                       :31:31+00:00   :31:40+00:00   :31:40+00:00
    1  DIR   ..            2023-12-13T15:25:18+00:  2024-03-01T11  2024-03-01T11  2024-03-01T11
                           00                       :31:31+00:00   :31:40+00:00   :31:40+00:00
    2  FILE  smb_test.txt  2024-03-01T11:21:23+00:  2024-03-01T11  2024-03-01T11  2024-03-01T11  5
                           00                       :21:43+00:00   :21:43+00:00   :21:43+00:00
```

**Comment by cgranleese-r7** [comment]
Tested `ls` changes to now take a path. As well as larger uploads and downloads.

## `ls <path>`
image

## `upload`/`download` larger files
**_Testing is a WIP_** I'll update this section once I have completed the testing.

## `<cmd> -h`
All commands now return the appropriate help output.

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 18]
```diff
@@ -318,7 +360,205 @@ def cmd_cat(*args)
...
+
+            remote_path = nil
+
+            @@mkdir_opts.parse(args) do |_opt, idx, val|
+              case idx
+              when 0
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_mkdir_help
```
Comment (adfoster-r7) [inline lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:505]:
```suggestion
                cmd_mkdir_help
                return
```

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb [hunk 19]
```diff
@@ -318,7 +344,198 @@ def cmd_cat(*args)
...
+                remote_path = val
+              else
+                print_warning('Too many parameters')
+                cmd_delete_help
+              end
+            end
+
+            full_path = as_ntpath(Pathname.new(shell.cwd).join(remote_path).to_s)
+            fd = simple_client.open(full_path, 'o')
+            fd.delete
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:478]: Can we `ensure` this happens 👀 

Side note; I wish the simple client offered a `proc` based API so all this cleanup would happen automagically

File: spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb (hunk 13 above)
Comment (adfoster-r7) [reply spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb:20]: Was this cromulent in the end? 👀

**Review by cgranleese-r7 — DISMISSED** [review]
Apart from some methods needing yard docs, it looks good to me 👍

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb (hunk 16 above)
Comment (adfoster-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:414]: I'll resolve this for now, but I think there's still an issue here

**Comment by cgranleese-r7** [comment]
Upload/download completion status:
image

File: lib/rex/post/smb/ui/console/command_dispatcher/shares.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply lib/rex/post/smb/ui/console/command_dispatcher/shares.rb:481]: `close` is the first thing that happens when calling off to `delete` https://github.com/rapid7/metasploit-framework/blob/3671f9dd6a37af2ab896254eae8bc2ebb81479a8/lib/rex/proto/smb/simple_client/open_file.rb#L19-L25

File: spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb (hunk 13 above)
Comment (dwelch-r7) [reply spec/lib/rex/post/smb/ui/console/command_dispatcher/core_spec.rb:20]: yup, sorry missed this somehow, works just fine

**Comment by adfoster-r7** [comment]
@dwelch-r7 Needs a rebase :+1:

**Comment by adfoster-r7** [comment]
## Release Notes

This PR adds the ability to upload/download/delete/mkdir/rmdir from within the SMB session type

**Pull request merged**